src/core/thread/mlr_manager.cpp

```diff
@@ -0,0 +1,353 @@
+    {
+        uint8_t pbbrServiceId;
+
+        SuccessOrExit(error = Get<BackboneRouter::Leader>().GetServiceId(pbbrServiceId));
+        SuccessOrExit(error = mle.GetServiceAloc(pbbrServiceId, messageInfo.GetPeerAddr()));
+    }
+    else
+    {
+        messageInfo.GetPeerAddr().SetToRoutingLocator(mle.GetMeshLocalPrefix(),
+                                                      Get<BackboneRouter::Leader>().GetServer16());
+    }
+
+    messageInfo.SetPeerPort(kCoapUdpPort);
+    messageInfo.SetSockAddr(mle.GetMeshLocal16());
+
+    SuccessOrExit(error = Get<Coap::Coap>().SendMessage(
+                      *message, messageInfo, &MlrManager::HandleMulticastListenerRegistrationResponse, this));
+
+exit:
+    if (error == OT_ERROR_NONE)
+    {
+        mMlrPending = true;
+        mTable.SetRegistering(num);
+    }
+    else
+    {
+        if (message != nullptr)
+        {
+            message->Free();
+        }
+
+        if (error == OT_ERROR_NO_BUFS)
+        {
+            ScheduleSend(1);
+        }
+    }
+
+    otLogInfoBbr("Send MLR.req: %s", otThreadErrorToString(error));
+    LogMulticastRegistrationTable();
+}
+
+void MlrManager::HandleMulticastListenerRegistrationResponse(Coap::Message *         aMessage,
+                                                             const Ip6::MessageInfo *aMessageInfo,
+                                                             otError                 aResult)
+{
+    uint8_t status = ThreadStatusTlv::MlrStatus::kMlrGeneralFailure;
+    otError error;
+
+    OT_UNUSED_VARIABLE(aMessageInfo);
+
+    mMlrPending = false;
+
+    VerifyOrExit(aResult == OT_ERROR_NONE && aMessage != nullptr, error = OT_ERROR_PARSE);
+    VerifyOrExit(aMessage->GetCode() == OT_COAP_CODE_CHANGED, error = OT_ERROR_PARSE);
+
+    SuccessOrExit(error = Tlv::FindUint8Tlv(*aMessage, ThreadTlv::kStatus, status));
+
+exit:
+
+    if (status == ThreadStatusTlv::MlrStatus::kMlrSuccess)
+    {
+        mTable.FinishRegistering(true);
+        // keep sending until all multicast addresses are registered.
+        ScheduleSend(0);
+    }
+    else
+    {
+        // Conservatively schedule reregistration if failed.
+        // This is required by Test Spec. MATN-TC-26
+        // TODO (MLR): reregister only failed addresses.
+        mTable.FinishRegistering(false);
+        UpdateReregistrationDelay(true);
+    }
+
+    otLogInfoBbr("Receive MLR.rsp, result=%s, status=%d, error=%s", otThreadErrorToString(aResult), status,
+                 otThreadErrorToString(error));
+}
+
+void MlrManager::HandleTimer(void)
+{
+    if (mSendDelay > 0)
+    {
+        if (--mSendDelay == 0)
+        {
+            SendMulticastListenerRegistration();
+        }
+    }
+
+    if (mReregistrationDelay > 0)
+    {
+        if (--mReregistrationDelay == 0)
+        {
+            Reregistration();
+        }
+    }
+
+    ResetTimer();
+}
+
+void MlrManager::Reregistration(void)
```

Comment:
nit: prefer to start method names with a verb - how about something like:
```suggestion
void MlrManager::Reregister(void)
```